### PR TITLE
Handle media player network errors and try to recover

### LINKED ISSUE #425: Handling media element errors
We added track error handling on #352 however [media element errors](http://w3c.github.io/html/semantics-embedded-content.html#error-codes) aren't being handled, which isn't a big deal since you can create your own media element with `createMediaElement` prop, however I think the library might benefit handling one error that can easily happen and gives a bad user experience, which is that when internet connection isn't very stable, the player might not play again until you load another track.

Steps to reproduce in Chrome (mobile/desktop):
- Play a track on https://benwiley4000.github.io/cassette/styleguide/.
- Take your browser offline and wait until no more requests are made (or media element throws an error).
- Get back online and try to play, and track won't play again.
- The media element should have the following error: `{code: 2, message: "PIPELINE_ERROR_READ: FFmpegDemuxer: data source error"}`.

Possible solutions within library:
- Adding a `onPlaybackFailure` prop, something similar to `onTrackPlaybackFailure` but for media element errors, and making `mediaCannotPlay` true or another boolean if it's more appropriate.
- Automatically trying to get media element out of the error state by reloading the track. Since we support passing a `startingTime`, if a network error is thrown for example, we could reload and seek to the time you were on as if nothing had happened, however `startingTime` is set in the playlist, so I'm not sure that can be used. This also bring questions like which errors would try to be solved by reloading, should max retries be set and others you might have.

Would like to know your thoughts and can help you making the PR if we agree to do something!

### PR DESCRIPTION
**Pull request by danielr18**
Fixes #425.

The logic is that when the player gets a network error, it checks if it's online to try and reload the active track, and if it's not, it listens to the [online event](https://developer.mozilla.org/en-US/docs/Web/API/NavigatorOnLine/Online_and_offline_events) and once the player is back online, it will try to reload.

The reload track method has an option to auto play the track, however I'm not sure if we should auto play after the error, the player becomes paused after the error, maybe it's best to just reload and leave it paused? Might be weird if you were offline for a considerate amount of time and then when you go back online, something starts playing.

The `online` event listener is removed:
- The event handler is called.
- The active track changes.
- The component unmounts.

@benwiley4000 Let me know what you think.

### CODE REVIEW

**Comment by danielr18** [comment]
I believe the `reloadActiveTrack` could be put in the control props. It would allow users to add a reload button when `mediaCannotPlay` is set. If you are offline and try to play a track, `mediaCannotPlay` will be true, and you can't play it again without changing tracks even after you come back online.

File: packages/core/src/PlayerContextProvider.js [hunk 1]
```diff
@@ -591,10 +601,41 @@ export class PlayerContextProvider extends Component {
...
+
   handleTrackPlaybackFailure(event) {
+    const shouldPlay = !this.state.paused;
     this.setState({
       mediaCannotPlay: true
     });
+    const error = event.target.error;
+    // If there's a network error, the current track should be reloaded
+    // in order to recover from the error state the player is in.
+    if (event.target === this.media && error && error.code === 2) {
```
Comment (benwiley4000) [inline packages/core/src/PlayerContextProvider.js]: Can we have a constant for the error code somewhere so we know which error type `2` is referencing?

**Review by benwiley4000 — CHANGES_REQUESTED** [review]
This is looking really good, thank you! And sorry for the delay in reviewing this.

> The reload track method has an option to auto play the track, however I'm not sure if we should auto play after the error, the player becomes paused after the error, maybe it's best to just reload and leave it paused? Might be weird if you were offline for a considerate amount of time and then when you go back online, something starts playing.

> I believe the `reloadActiveTrack` could be put in the control props. It would allow users to add a reload button when mediaCannotPlay is set. If you are offline and try to play a track, mediaCannotPlay will be true, and you can't play it again without changing tracks even after you come back online.

Both great points! I think if we include this control prop then it would obviate the need for the autoplay behavior. I suggest we go with that!

**Comment by danielr18** [comment]
Just made the changes, feel free to modify the documentation of the new control prop if you want to improve it.

**Comment by benwiley4000** [comment]
Great thank you! I'll try to release this soon.

**Comment by benwiley4000** [comment]
@danielr18 released in v2.0.0-beta.2!

**Pull request merged**